### DISCUSSION TITLE
unable to install mesa

### DISCUSSION THREAD

**Discussion by JilinHu** [Homebrew/discussions#2787] [Linux] [answered]
Hi,  While I was trying to install mesa through brew command on Ubuntu 20.04, it always fails and shows the error below. I have reinstalled libdrm or llvm and so on but it doesn't work. Anybody knows? Thanks a lot.

![unnamed](https://user-images.githubusercontent.com/40224482/149456849-700f16a9-4e5f-428b-9bf5-f10e27718488.png)
![unnamed](https://user-images.githubusercontent.com/40224482/149456862-7f6c0caf-f53f-487e-ad46-005c7e86f0ff.png)

**Comment by carlocab** [comment]
Please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse (e.g. when looking up error messages). Instead, copy the text here and wrap them in three back-ticks (```) on the line before and on the line after to format them properly.

What is the output of `brew config` and `brew doctor`? Also, please do `brew gist-logs mesa` and post the resulting link here.

**Reply by JilinHu** [reply]
Thanks for your kind remind! I'll pay attention to and correct it.
The result of  'brew config' is as below.
```
hujilin@DESKTOP-JKPTATT:~$ brew config
HOMEBREW_VERSION: 3.3.9-198-g385892f
ORIGIN: https://github.com/Homebrew/brew
HEAD: 385892f3d27dcab0f5a1cb47c3927d9d65ed007c
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core.git
Core tap HEAD: 2997a91345e68ea724ec6160e39d435d2a48df2e
Core tap last commit: 3 days ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: 16-core 64-bit unknown_0x6_0xa5
Clang: 13.0.0
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /bin/curl
Kernel: Linux 4.4.0-19041-Microsoft x86_64 GNU/Linux
OS: Ubuntu 20.04.3 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.4.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@5: N/A
xorg: N/A
```

And the result of 'brew doctor' is here. As I installed homebrew manually, so I thought it's not a problem as I can use brew to install other formulaes.
```
hujilin@DESKTOP-JKPTATT:~$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /home/linuxbrew/.linuxbrew.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/home/linuxbrew/.linuxbrew).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.
```

The resulting link of 'brew  gist-logs mesa' is:
```
https://gist.github.com/8e791b2beb9fff9089f69a37a7b4a488
```
Thanks again.

**Reply by gromgit** [reply]
You've deviated from a standard Homebrew Linux installation in at least two problematic ways:
1. You're using a non-standard prefix (`/usr/local` instead of `/home/linuxbrew/.linuxbrew`).
2. You're running WSL instead of "native" Linux.

If you're running WSL1 instead of WSL2, that's yet another point of deviation. As far as I know, source builds on user installations have never been guaranteed to work, WSL installations doubly so. There are simply too many things that can go wrong, and the Mesa build logs you posted don't really shed any light (at least to me) about the root cause of your build failure.

Unless you absolutely cannot do so, you really should install Homebrew under `/home/linuxbrew/.linuxbrew` _and_ under WSL2 (if you aren't doing so already). That means your formula installs will largely use pre-built binaries, which is one HUGE variable removed from any further troubleshooting that may be required.

**Reply by carlocab** [reply]
Yes, if you're on WSL1 that might be this problem: #2777

**Reply by JilinHu** [reply]
Thank you! My wsl version  is 2 but I am not going to continue on this system. Instead I install ubuntu and try to install brew through command on hoomebrew page. But I received an error which I met sometimes and I dont know how to solve it.
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"


==> Tapping homebrew/core
remote: Enumerating objects: 1127240, done.
remote: Counting objects: 100% (49/49), done.
remote: Compressing objects: 100% (35/35), done.
error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
fatal: the remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
Failed during: git fetch --force origin refs/heads/master:refs/remotes/origin/master

```
I have tryed the command below but it dosn't work.
```
git config --global http.postBuffer 524288000
ifconfig ethn0 mtu 1400
```

**Reply by gromgit** [reply]
>```
> error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
>```
That's a network failure, but at least you got past the Homebrew software installation part. Try the following:
```sh
# set up Homebrew environment
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
brew update-reset

# if the previous command completed successfully...
brew install mesa
```
Once you've verified that Homebrew works as expected, you probably want to add the `eval` command above at the end of the startup script for your login shell. For bash, it's `~/.bash_profile` or `~/.profile`, whichever exists. For Zsh, it's `~/.zprofile`.

**Reply by JilinHu** [reply]
Well I have encountered another problem while excuting brew fetch llvm. It always got stuck at 42.0% and the total file is about 350MB,
unfortunately I didn't find useful solution on the Internet yet. I thought it's a network failure but don't know how to fix it. 
```
brew fetch llvm

==> Downloading https://ghcr.io/v2/homebrew/core/llvm/manifests/13.0.0_2
Already downloaded: /home/hujilin/.cache/Homebrew/downloads/489f0c3e95a69b835d9bbbe1d7099efb19b5690c2dca67e86e7a1f7d502a33e9--llvm-13.0.0_2.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/llvm/blobs/sha256:d4a9f80cc2318943e7526808506203ee104af83682e447be00c69243643401c9
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:d4a9f80cc2318943e7526808506203ee104af83682e447be00c69243643401c9?se=2022-01
##############################                                            41.9%curl: (18) transfer closed with 213333556 bytes remaining to read

Error: Failed to download resource "llvm"
Download failed: https://ghcr.io/v2/homebrew/core/llvm/blobs/sha256:d4a9f80cc2318943e7526808506203ee104af83682e447be00c69243643401c9
Warning: Bottle fetch failed, fetching the source instead.
==> Downloading https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/llvm-project-13.0.0.src.tar.xz
Already downloaded: /home/hujilin/.cache/Homebrew/downloads/8fd68fc8f968137c5080826db6e58682326235960fd8469363eb27d0799978ca--llvm-project-13.0.0.src.tar.xz
SHA256: 6075ad30f1ac0e15f07c1bf062c1e1268c241d674f11bd32cdf0e040c71f2bf3

```
brew doctor
```

```

brew config
```
HOMEBREW_VERSION: 3.3.10-58-g069ab08
ORIGIN: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git
HEAD: 069ab087f99fd183aee599bf785cb835d5868407
Last commit: 12 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 5130bfe526e1a959c0ba889785b17d1da5da238d
Core tap last commit: 4 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew/Homebrew
HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 16
all_proxy: socks://127.0.0.1:7891/
http_proxy: http://127.0.0.1:7890
https_proxy: http://127.0.0.1:7890
no_proxy: localhost,127.0.0.1,localaddress,.localdomain.com
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: 16-core 64-bit unknown_0x6_0xa5
Clang: N/A
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /bin/curl
Kernel: Linux 5.11.0-46-generic x86_64 GNU/Linux
OS: Ubuntu 20.04.3 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.3.0
/usr/bin/ruby: 2.7.0
glibc: N/A
gcc@5: N/A
xorg: N/A

```

**Comment by gromgit** [comment]
>```
>ORIGIN: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git
>```
It appears you're behind the Great Firewall of China, so you may be better served by a local bottle mirror instead. See https://github.com/Homebrew/discussions/discussions/1917 for details.

**Reply by JilinHu** [reply]
Well Thanks for your reply. That's true. And It's really weird that I download it successfully just switching to a high speed network.

**Reply by gromgit** [reply]
Not weird at all. That's probably why your compatriots set up the mirrors in the first place, because large downloads through a national firewall tend to break.

**Reply by JilinHu** [reply]
After setting this. I can install the example hello, but I got stuck at ghostscript. This man had the problem alike(https://github.com/Homebrew/homebrew-core/issues/88617)

While I try : brew fetch --force --debug --verbose I got the response below. 
```
==> Downloading http://git.ghostscript.com/?p=ghostpdl.git;a=patch;h=830afae5454dea3bff903869d82022306890a96c
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Linuxbrew/3.3.10-58-g069ab08\ \(Linux\;\ x86_64\ Ubuntu\ 20.04.3\ LTS\)\ curl/7.68.0 --header Accept-Language:\ en --retry 3 --location --head http://git.ghostscript.com/\?p=ghostpdl.git\;a=patch\;h=830afae5454dea3bff903869d82022306890a96c
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Linuxbrew/3.3.10-58-g069ab08\ \(Linux\;\ x86_64\ Ubuntu\ 20.04.3\ LTS\)\ curl/7.68.0 --header Accept-Language:\ en --fail --retry 3 --location --remote-time --output /home/hujilin/.cache/Homebrew/downloads/e76c7e0ae52dbd20f28f187e9cb784e87ef49e8e4b35de8635499754be1d0d6c--ghostpdl.git-830afae5454dea3bff903869d82022306890a96c.patch.incomplete http://git.ghostscript.com/\?p=ghostpdl.git\;a=patch\;h=830afae5454dea3bff903869d82022306890a96c
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   809    0   809    0     0      3      0 --:--:--  0:03:44 --:--:--     0
```

It just stop download here.
```
Downloading http://git.ghostscript.com/?p=ghostpdl.git;a=patch;h=830afae5454dea3bff903869d82022306890a96c
```

**Reply by gromgit** [reply]
> While I try : brew fetch --force --debug --verbose [...] It just stop download here.

I can confirm that patch being downloaded is exactly 809 bytes long, so it looks like the Great Firewall is refusing to close your connection for some reason. The folks in a position to help resolve your issue are on _your_ side of the firewall.

That said, as long as you're using an up-to-date core tap (and your `brew config` says you are), `brew fetch` should be pulling down a bottle, not the source code as your output indicates. What's the output of `brew info ghostscript`?

**Comment by JilinHu** [comment]
I just delete these lines cause it's the reson why I can't continue but it shows another issues so I thought the patch is also necessary for building.
```
patch do
   url "http://git.ghostscript.com/?p=ghostpdl.git;a=patch;h=830afae5454dea3bff903869d82022306890a96c"
    sha256 "9c282e3f732b5c182ab761e187850d9f4c29d938a93e3785d9cb6ced13111d6c"
  end

```
```
==> ./configure --prefix=/home/linuxbrew/.linuxbrew/Homebrew/Cellar/ghostscript/9.55.0 --libdir=/home/linuxbrew/.linuxbrew/H
==> make install
Last 15 lines from /home/hujilin/.cache/Homebrew/Logs/ghostscript/02.make:
DEVICE_DEVS_EXTRA= \
sh <./obj/xpsldt.tr
/home/linuxbrew/.linuxbrew/Homebrew/bin/ld: ./obj/gsicc_lcms2.o: in function `gscms_transform_color':
gsicc_lcms2.c:(.text+0x6a3): undefined reference to `gscms_transformm_color_const'
/home/linuxbrew/.linuxbrew/Homebrew/bin/ld: ./obj/gsicc_lcms2.o: in function `gscms_transform_color':
gsicc_lcms2.c:(.text+0x6a3): undefined reference to `gscms_transformm_color_const'
collect2: error: ld returned 1 exit status
make: *** [base/unixlink.mak:196: bin/gpcl6] Error 1
make: *** Waiting for unfinished jobs....
collect2: error: ld returned 1 exit status
make: *** [base/unixlink.mak:172: bin/gs] Error 1
/home/linuxbrew/.linuxbrew/Homebrew/bin/ld: ./obj/gsicc_lcms2.o: in function `gscms_transform_color':
gsicc_lcms2.c:(.text+0x6a3): undefined reference to `gscms_transformm_color_const'
collect2: error: ld returned 1 exit status
make: *** [base/unixlink.mak:217: bin/gxps] Error 1
```
The output of bew info ghostscript is:
```
ghostscript: stable 9.55.0, HEAD
Interpreter for PostScript and PDF
https://www.ghostscript.com/
Not installed
From: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git/Formula/ghostscript.rb
License: AGPL-3.0-or-later
==> Dependencies
Build: pkg-config ✔
Required: fontconfig ✔, freetype ✔, jbig2dec ✔, jpeg ✔, libidn ✔, libpng ✔, libtiff ✔, little-cms2 ✔, openjpeg ✔, expat ✔, zlib ✔, gcc ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 55,947 (30 days), 259,555 (90 days), 966,652 (365 days)
install-on-request: 11,275 (30 days), 47,614 (90 days), 170,492 (365 days)
build-error: 19 (30 days)

```

**Reply by carlocab** [reply]
Not sure why your `ghostscript` bottle is missing. You shouldn't need to build this from source. In any case, it might help to unset these
```
all_proxy: socks://127.0.0.1:7891/
http_proxy: http://127.0.0.1:7890
https_proxy: http://127.0.0.1:7890
no_proxy: localhost,127.0.0.1,localaddress,.localdomain.com
```
and then try `brew fetch ghostscript`. If that doesn't work, try `brew fetch --build-from-source ghostscript`.

**Reply by gromgit** [reply]
> Not sure why your `ghostscript` bottle is missing.

I suspect this:
>```
>From: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git/Formula/ghostscript.rb
>```
has something to do with it.

@asdasdssw, you might want to ask the Tsinghua mirror maintainers why their `ghostscript` mirror formula has no bottles.

**Reply by JilinHu** [reply]
yeh, it's funny, I am doubting it too so I change it to the original bottle on github. But it succeed after I unset the proxy rather than before, it's confusing.

**Reply by carlocab** [reply]
> > Not sure why your `ghostscript` bottle is missing.
> 
> I suspect this:
> 
> > ```
> > From: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git/Formula/ghostscript.rb
> > ```
> 
> has something to do with it.

Probably, but I find that surprising, since
```
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
```

It looks like the Tsinghua mirror is actually a fork and not a mirror. I don't blame them, since changes in `brew` seem to break them all the time.

**Reply by carlocab** [reply]
On the other hand, this commit checks out:
```
ORIGIN: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git
HEAD: 069ab087f99fd183aee599bf785cb835d5868407
```
So... not sure why the formula is still being pulled from a mirror. Proxy settings still maybe?

**Comment by JilinHu** [comment]
Thank you. Unseting the proxy works!  But I don't know why as the proxy helps me to access to outer-Internet. I have set the homebrew ORIGIN and bottle to the real github rather than the mirrors. In that case I shouldn't access the outer-Internet as the proxy is off.
By the way, why the proxy lines still shows below after I unset them?

Here is the result of brew config
```
HOMEBREW_VERSION: 3.3.10-58-g069ab08
ORIGIN: https://github.com/Homebrew/brew
HEAD: 069ab087f99fd183aee599bf785cb835d5868407
Last commit: 17 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0b420a996230eb5aef62a858c93741839aa7c9bc
Core tap last commit: 85 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew/Homebrew
HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 16
all_proxy: socks://127.0.0.1:7891/
http_proxy: http://127.0.0.1:7890
https_proxy: http://127.0.0.1:7890
no_proxy: localhost,127.0.0.1,localaddress,.localdomain.com
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: 16-core 64-bit unknown_0x6_0xa5
Clang: N/A
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /bin/curl
Kernel: Linux 5.11.0-46-generic x86_64 GNU/Linux
OS: Ubuntu 20.04.3 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.3.0
/usr/bin/ruby: 2.7.0
glibc: N/A
gcc@5: N/A
xorg: N/A

```

brew info ghostscript again
```
hostscript: stable 9.55.0, HEAD
Interpreter for PostScript and PDF
https://www.ghostscript.com/
/home/linuxbrew/.linuxbrew/Homebrew/Cellar/ghostscript/9.55.0 (683 files, 169.9MB) *
  Built from source on 2022-01-15 at 20:34:14
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/ghostscript.rb
License: AGPL-3.0-or-later
==> Dependencies
Build: pkg-config ✔
Required: fontconfig ✔, freetype ✔, jbig2dec ✔, jpeg ✔, libidn ✔, libpng ✔, libtiff ✔, little-cms2 ✔, openjpeg ✔, expat ✔, zlib ✔, gcc ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 55,947 (30 days), 259,555 (90 days), 966,652 (365 days)
install-on-request: 11,275 (30 days), 47,614 (90 days), 170,492 (365 days)
build-error: 19 (30 days)

```

**Reply by gromgit** [reply]
Something is still wrong.
>```
>ghostscript: stable 9.55.0, HEAD
>```
should be:
>```
>ghostscript: stable 9.55.0 (bottled), HEAD
>```

> By the way, why the proxy lines still shows below after I unset them?

Because you didn't unset them in the environment you ran `brew config`. `unset all_proxy http_proxy https_proxy no_proxy`, then `brew config` again to confirm that they're really unset, then do all other Homebrew stuff in this shell.

Once you've confirmed that you really have unset those variables, `brew update-reset` to reset your core tap, then `brew info ghostscript` again. If you still don't see `(bottled)`, your Homebrew setup is pretty badly messed up.

**Comment by JilinHu** [comment, marked as answer]
Thanks for your suggestions!  I'll follow them later and now I get back to the origin problem with mesa. 
```
/home/linuxbrew/.linuxbrew/Homebrew/bin/ld: src/gallium/drivers/r600/libr600.a: error adding symbols: malformed archive
collect2: error: ld returned 1 exit status
[2502/2503] Linking target src/intel/vulkan/libvulkan_intel.so
[2503/2503] Linking target src/mesa/drivers/dri/libmesa_dri_drivers.so
ninja: build stopped: subcommand failed.
```

And here is the result of brew gist-logs mesa
https://gist.github.com/16844d4333cc602e80931daf6a686bdb, hope something changes now.

**Reply by gromgit** [reply]
I don't see any difference from your first run, but I confirmed that `mesa` rebuilds correctly on both my own Ubuntu 21.10 setup and Homebrew's own [Ubuntu 20.10 Docker container](https://github.com/Linuxbrew/docker), so it's likely to be something unique to your setup.

Speaking of setup, this part of your `brew config` output is troubling:
>```
>HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Homebrew/Cellar
>```
Your non-standard Cellar location is the reason why you're forced to rebuild `mesa`; it and a couple thousand other formulae have _non-relocatable_ bottles that absolutely must reside in `/home/linuxbrew/.linuxbrew/Cellar`. Unless you plan to spend days trying to figure out what's happening, I strongly recommend using the official Homebrew installer, which puts things in the Right Places and lets you use bottles to the full.

**Reply by JilinHu** [reply]
Thank you very much! I finally make it. I didn't know that the prefix path really matter a lot. And I also got messed up with the environment variable. So I uninstall the homebrew, and reinstall it following the guide in my country's mirror website step by step. Then I get formulaes downloads with bottle rather than source and it becomes faster to install them.
I really learned a lot through our discussion and thank you again.@gromgit@carlocab
